reconcile/deadmanssnitch.py

```diff
@@ -0,0 +1,244 @@
+    DeadMansSnitchSettingsV1,
+)
+from reconcile.gql_definitions.common.clusters_with_dms import ClusterV1
+from reconcile.typed_queries.app_interface_deadmanssnitch_settings import (
+    get_deadmanssnitch_settings,
+)
+from reconcile.typed_queries.clusters_with_dms import get_clusters_with_dms
+from reconcile.utils.deadmanssnitch_api import (
+    DeadMansSnitchApi,
+    Snitch,
+)
+from reconcile.utils.runtime.integration import (
+    NoParams,
+    QontractReconcileIntegration,
+)
+from reconcile.utils.secret_reader import (
+    SecretNotFound,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.vault import (
+    VaultClient,
+    _VaultClient,
+)
+
+QONTRACT_INTEGRATION = "deadmanssnitch"
+SECRET_NOT_FOUND = "SECRET_NOT_FOUND"
+
+
+class Action(enum.Enum):
+    create_snitch = "create_snitch_url"
+    delete_snitch = "delete_snitch_url"
+    update_vault = "update_vault_data"
+
+
+class DiffData(BaseModel, ABC):
+    """Base class to store the reconcile activity on current state derived from desired state"""
+
+    cluster_name: str
+    action: Action
+
+
+class CreateSnitchDiffData(DiffData):
+    """Class to store data used to create snitch"""
+
+    action = Action.create_snitch
+
+
+class UpdateVaultDiffData(DiffData):
+    """Class to store data used to update vault with snitch url"""
+
+    action = Action.update_vault
+    check_in_url: str
+
+
+class DeleteSnitchDiffData(DiffData):
+    """Class to store data used to delete snitch"""
+
+    action = Action.delete_snitch
+    token: str
+
+
+class DiffHandler:
+    """Handler class which to create/delete/update vault based on action on DiffData"""
+
+    def __init__(
+        self,
+        deadmanssnitch_api: DeadMansSnitchApi,
+        settings: DeadMansSnitchSettingsV1,
+        vault_client: _VaultClient,
+    ) -> None:
+        self.deadmanssnitch_api = deadmanssnitch_api
+        self.settings = settings
+        self.vault_client = vault_client
+
+    def summarize(self, diffs: Sequence[DiffData]) -> str:
+        return "\n".join(
+            f"cluster name: {diff.cluster_name} - action: {diff.action.value}"
+            for diff in diffs
+        )
+
+    def apply_diff(self, diff: DiffData) -> None:
+        match diff:
+            case CreateSnitchDiffData(cluster_name=cluster_name):
+                self.create_snitch(cluster_name)
+            case DeleteSnitchDiffData(cluster_name=cluster_name, token=token):
+                self.deadmanssnitch_api.delete_snitch(token)
+            case UpdateVaultDiffData(
+                cluster_name=cluster_name, check_in_url=check_in_url
+            ):
+                self.vault_client.write(
+                    {
+                        "path": self.settings.snitches_path,
+                        "data": {f"deadmanssnitch-{cluster_name}-url": check_in_url},
+                    },
+                    decode_base64=False,
+                )
```

Comment:
just my opinion, but i feel that this class is only causing us to move data from place to place, and the actual functionality can be significantly reduced and added into `DeadMansSnitchIntegration`.